hts_expr.c

```diff
@@ -377,12 +377,15 @@ static int unary_expr(hts_filter_t *filt, void *data, hts_expr_sym_func *fn,
         }
     } else if (*str == '!') {
         err = unary_expr(filt, data, fn, str+1, end, res);
-        if (!hts_expr_val_exists(res)) {
-            // We can still negate undef values by toggling the
+        res->is_str = 0;
+        if (res->is_true) {
+            // Any explicitly true value becomes false
+            res->d = res->is_true = 0;
+        } else if (!hts_expr_val_exists(res)) {
+            // We can also still negate undef values by toggling the
             // is_true override value.
             res->d = res->is_true = !res->is_true;
         } else if (res->is_str) {
-            res->is_str = 0;
```

Comment:
It's looking better, but there's a small problem with this change.  Setting `res->is_str = 0;` before all the tests means `res->is_str` here will always be false...